nx_cugraph/convert_matrix.py

```diff
@@ -276,3 +284,95 @@ def from_scipy_sparse_array(
     if inplace:
         return create_using._become(G)
     return G
+
+
+@networkx_algorithm(version_added="25.06", fallback=True)
+def to_numpy_array(
+    G,
+    nodelist=None,
+    dtype=None,
+    order=None,
+    multigraph_weight=sum,
+    weight="weight",
+    nonedge=0.0,
+):
+    """MultiGraphs are not yet supported"""
+    # print(" ==> hello!! dispatched to nxcg!")
+    G = _to_graph(G, weight, 1, _get_float_dtype(dtype))
+
+    if nodelist is None:
+        nodelist = list(G)
+
+    N = len(nodelist)
+
+    # use set to check for nodes not in the graph or duplicate nodes
+    nodelist_as_set = set(nodelist)
+    if nodelist_as_set - set(G):
+        raise nx.NetworkXError(
+            f"Nodes {nodelist_as_set - set(G)} in nodelist is not in G"
+        )
+    if len(nodelist_as_set) < N:
+        raise nx.NetworkXError(f"Nodelist {nodelist} contains duplicates")
+
+    # Construct array of fill_value matching resulting shape
+    A = cp.full((N, N), fill_value=nonedge, dtype=dtype, order=order)
+
+    # Case: empty nodelist or graph without any edges
+    if N == 0 or G.number_of_edges() == 0:
+        return A
+
+    # If dtype is structured and weight is None, use dtype field names as
+    # edge attribs
+    edge_attrs = None
+    if A.dtype.names:
+        if weight is None:
+            edge_attrs = dtype.names
+        else:
+            raise ValueError(
+                "Specifying `weight` not supported for structured dtypes\n."
+                "To create adjacency matrices from structured dtypes,"
+                "use `weight=None`."
+            )
+
+    # Map nodes to row/col in matrix
+    # idx = dict(zip(nodelist, range(N)))
+    if len(nodelist) < len(G):
+        G = G.subgraph(nodelist)
+
+    # Collect all edge weights and reduce with `multigraph_weights`
+    # TODO handle this case
+    if G.is_multigraph():
+        if edge_attrs:
+            raise nx.NetworkXError(
+                "Structured arrays are not supported for MultiGraphs"
+            )
+    else:
+        # Special case: TODO: figure out
+        if edge_attrs:
+            pass
```

Comment:
I was playing around with this case (trying to get cupy to work), and here's what I came up with (there may be better ways):
```python
            edge_array = np.empty(src_indices.size, dtype=dtype)
            for edge_attr in edge_attrs:
                if edge_attr in G.edge_values:
                    e_array = G.edge_values[edge_attr]
                    if edge_attr in G.edge_masks:
                        e_array = cp.where(G.edge_masks[edge_attr], e_array, 1)
                else:
                    e_array = ...  # maybe np.ones?
                edge_array[edge_attr] = cp.asnumpy(e_array)
```